### ISSUE TITLE
Support for AlpineJS in slots

### ISSUE THREAD

**Issue by JuroOravec** [EmilStenstrom/django-components#644]
### Background

When I have a slot with `isolated` context, I expect that the variables accessible inside the slot are the same as outside of the component. In other words, whatever variables the component defines, it does NOT leak into the slot, unless explicitly set:

```django
{{ outside_var }}
{% component "table" %}
	{{ outside_var }}    <-- same as above
{% endcomponent %}
```

For the UI component library, I'm using AlpineJS. To emulate slot Vue behaviour, I need this behavior of isolated context to be true also for AlpineJS variables:

```django
<div x-data="{ outsideVar: 123 }">
  {% component "table" %}
    <div x-text="outsideVar"> <-- same as above
	</div>
  {% endcomponent %}
</div>
```

### Implementation

I was able to achieve this using [x-teleport](https://alpinejs.dev/directives/teleport) like so:
1. Django component that supports these sort of "alpine-enabled" slots defines the django slots either at the very end or very start of its template, so they are not nested in anything else.
2. But instead of using plain `{% slot %}` tag, I wrap it in two HTML elements, `<template>` and `<div>`
3. And then, at the place where I wanted to ORIGINALLY put the slot, I define an element with an ID into which the slot is inserted

```django
<template x-teleport="#abc">
    <div {% html_attrs attrs %}>
        {% slot "default" default %}
    </div>
</template>

{# The actual component body #}
<main>
	<div>
        <span id="abc">
		</span>
        ...
	</div>
</main>
```

Notes:
- `<template>` is required by AlpineJS, and that's where we define the teleport
- `<div>` inside the `<template>` is used for 2 reasons:
	1. Because x-teleport requires a single element inside it, so by inserting the `<div>`, we allow the actual slot content to be anything.
	2. As a place to bind slot data to the slot, similarly how we do `{% slot data="var" %}`

### Accessing slot data

By adding `x-data` to the `<div>`, I am able to expose the Alpine data to the slot the same way I can expose Django data to the slot using `{% slot data="var" %}`:

```django
<template x-teleport="#abc">
    <div x-data="{ $slot: data }" {% html_attrs attrs %}>
        {% slot "default" default %}
    </div>
</template>
```

So, from within the slot, the data can then be accessed as `$slot`:

```django
<div x-data="{ outsideVar: 123 }">
  {% component "table" %}
    <div x-text="outsideVar"> <-- same as above
	</div>
    <div x-text="$slot.exposedVar"> <-- exposed from "table"
	</div>
  {% endcomponent %}
</div>
```

### API

Ideally, this would be marked on the `{% slot %}` tag. Plus we need to allow define what Alpine data should be exposed to the slot.

So it could look like this:
```django
{# Normal slot #}
{% slot "mytable" / %} 

{# Alpine slot, uses `x-teleport`, `$slot` is implicitly an empty JS object #}
{% slot "mytable" alpine / %}

{# `$slot` is explicitly an empty JS object #}
{% slot "mytable" alpine="{}" / %}

{# `$slot` is explicitly a JS object #}
{% slot "mytable" alpine="{ abc: 123 }" / %}

{# `$slot` is explicitly an Alpine variable myJsVar #}
{% slot "mytable" alpine="myJsVar" / %}
```

If the slot HAS an `alpine` keyword, we render the `<template x-teleport="#abc">` at the very start or an end of the django template. The `{% slot %}` would be inside the `<template>`. And, in place of the original slot position, we insert `<div id="abc"></div>`.

### Other considerations

- The generated IDs could look like "slot-{component_id}-{slot_name}", e.g. `slot-a0b1c3-content`

- Because we use the `<div>` inside the x-teleport, and another to specify the target ID, it means that slots using the `alpine` keyword would have 2 extra `<div>`s wrapping the actual content.

	So if I had
	
	```
	{% component "table" %}
	  {% slot "content" alpine %}
	    Hello
	  {% endslot %}
	{% endcomponent %}
	```
	
	In the slot's position, we'd actually render:
	```html
	<div id="abc">
	    <div x-data="{ $slot: data }">
			hello
		<div>
	</div>
	```

	This could break some styling setups, as users wouldn't be able to modify the divs.

	So instead we could add CSS classes to the slots:
	```html
	<div id="abc" class="slot-outer__my_slot">
	    <div x-data="{ $slot: data }" class="slot-inner__my_slot">
			hello
		<div>
	</div>
	```

	Then, together with [scoped CSS](https://github.com/EmilStenstrom/django-components/issues/622), people would be able to style these as:

	```
	me .slot-outer__my_slot {
		flex: auto;
	}

	me .slot-inner__my_slot {
		background: red;
	}
	```

**Comment by EmilStenstrom** [comment]
Not knowing AlpineJS, I'm trying to wrap my head around this. Are you saying that we should add support for AlpineJS to slot tags? I was hoping that we could avoid having any code that is specific to a certain library in django_components, and that this was something that anyone can add in a separate library? Are you thinking differently?

**Comment by JuroOravec** [comment]
I'm thinking the same, ideally it would be a plugin. But right now it's hard to imagine how plugins should look like, so I was thinking of first implementing this, and then reverse-engineering API for plugins. But we can also do it in a single step.

For the Alpine slots support, I imagine the plugin would:
- Pre-process inputs passed to tags, e.g. so slot could accept `alpine="{}"` keyword, which the plugin would detect, and remove from inputs.
- Once a tag was been processed into a Node, the plugin would then need to store the plugin-specific data somewhere. So each Node would have a `meta` or `extra` attribute, where plugin authors could store the extra data that could then later be accessed in render phase.

	So e.g. given `{% slot alpine="{ abc: 123 }" %}`, the plugin would extract and set `SlotNode.meta["alpine_slots__data"] = "{ abc: 123 }"`

	`alpine_slots__` being a unique prefix for the plugin.
- They could also modify the template (or rather, make copies from the original templates?)
  So that the support for Alpine slots could be implmenteded as template transformation.

**Comment by EmilStenstrom** [comment]
Not sure if this warrants the complexity of maintaining a whole plugin system, but you're doing the work here :)

**Comment by jdare-compass** [comment]
Just wanted to chime in that I'm running into this issue as well and would make use of a plugin that addresses it

**Comment by EmilStenstrom** [comment]
@jdare-compass @zachbellay Tell us more. In your own words: What is the problem, and how should we solve it in your opinion?

**Comment by JuroOravec** [comment]
## TL;DR - End result

So with the AlpineJS integration (details below), an example component could look like this:

In Python you simply prepare the data and return it from `get_js_data()`.

```py
from typing import NamedTuple
from typing_extensions import NotRequired, TypedDict
from django_components import Component


class MultiselectJsProps(TypedDict):
    selected_items: NotRequired[str]
    all_items: NotRequired[str]
    passthrough: NotRequired[str]


class Multiselect(Component):
   class Kwargs(NamedTuple):
       selected_items: list | None = None
       js: MultiselectJsProps | None = None

   def get_js_data(self, args, kwargs: Kwargs, slots, context):
       if kwargs.selected_items:
           selected_items = [
               SelectOption(value=item, label=item, attrs={})
               if not isinstance(item, SelectOption)
               else item
               for item in input_kwargs["selected_items"]
           ]
       elif kwargs.js.get("selected_items"):
           selected_items = js(kwargs.js)
       else:
           raise ValueError("Missing required kwarg 'selected_items'")

       return {
           "selectedItems": selected_items,
           "allItems": [...],
           # To set event listeners, use `on` + event name
           "onChange": js("() => console.log('Hello!')"),
       }
```

In JS export an object that defines the AlpineJS component. Here you can access the data from `get_js_data()` as "props" and set up watchers, callbacks, etc. The data returned from the `setup()` method will be available in the template as AlpineJS variables:

 ```js
  // Define component similarly to defining Vue components
  export default {
    props: {
      /* { label: string, value: string, attrs?: object }[] */
      allItems: { type: Array, required: true },
      selectedItems: { type: Array, required: true },
    },
  
    emits: {
      change: (payload) => true,
    },
  
    // Instead of Alpine's init(), use setup()
    // Props are passed down as reactive props, same as in Vue
    // Second argument is the Alpine component instance.
    // Third argument is the reactivity API, equivalent to `@vue/reactivity`
    setup(props, vm, { computed, ref, watch }) {
      // Variables
      const allItems = ref([]);
      const selectedItems = ref([]);
      const items = ref([]);
  
      // Computed
      const allItemsByValue = computed(() => {
        return allItems.value.reduce((acc, item) => {
          acc[item.value] = item;
          return acc;
        }, {});
      });
  
      // Set the initial state from HTML
      watch(() => props.allItems, () => {
        allItems.value = props.allItems;
      }, { immediate: true })
  
      watch(() => props.selectedItems, () => {
        selectedItems.value = props.selectedItems;
      }, { immediate: true })
  
      // Watch for changes
      watch(selectedItems, () => {
        onItemsChange();
      }, { immediate: true });
  
      // Methods
      const addItem = () => {
        const availableItems = getAvailableItems();
        if (!availableItems.length) return;
  
        // Add item by removing it from available items
        const nextValue = availableItems.shift();
        const newSelectedItems = [
          ...selectedItems.value,
          nextValue,
        ];
  
        // And add it to the selected items
        selectedItems.value = newSelectedItems;
      }
  
      // ...

      return {
        items,
        allItems,
        addItem,
      };
    },
  };
 ```

And in HTML you can access the fields returned from JS's `setup()` method:

```django
<div class="pt-3 flex flex-col gap-y-3 items-start">
  {% slot "title" / %}

  <template x-for="(item, index) in selectedItems.value" :key="item.value">
    <div class="inline-flex items-center w-full px-2.5 text-sm">
      <span x-html="item.content" class="w-full"></span>
      {% if editable %}
        {% component "Icon"
          name="x-mark"
          attrs:class="ml-1.5 hover:text-gray-400 text-gray-600 cursor-pointer focus:outline-none"
          attrs:@click.stop="removeItem(item)"
        / %}
      {% endif %}
    </div>
  </template>
  ...
</div>
```

## Observations

Few observations on integrating JS (AlpineJS) and Python (django-components) from my work project from the other day:

Note that  I'm using my [alpine-composition](https://github.com/JuroOravec/alpinui/tree/main/packages/alpine-composition) and [alpine-reactivity](https://github.com/JuroOravec/alpinui/tree/main/packages/alpine-reactivity) packages to be able to write JS logic similar to Vue.

1. The component tag `{% component %}` is THE interface for passing anything and everything to the component - Python variables, slots, but also JS variables.

   To distinguish JS from Python variables, I scope all the JS inputs under a `js` dictionary, which means that in the template, I prefix JS variables with `js:` (e.g. `js:onChange`):

   ```django
   <div class="flex gap-5">
     {% component "Multiselect"
       all_items=options
       selected_items=selected_items
       js:onChange="(items) => setAssignees(items)"
     / %}
   </div>
   ```

    Whereas the component class may look like this:

    ```py
    class MultiselectJsProps(TypedDict):
        selected_items: NotRequired[str]
        all_items: NotRequired[str]
        passthrough: NotRequired[str]

    class Multiselect(Component):
        class Kwargs(NamedTuple):
            all_items: list | None = None
            selected_items: list | None = None
            js: MultiselectJsProps | None = None
    ```

    All values of the `js` dict (`MultiselectJsProps`) are strings, because they are JS snippets ran within AlpineJS context.

2. As you can see above, the same kwargs, like `selected_items` can be passed either as Python variable or as JS variable. This means that:

   - Inside the `get_template_data()` of `Multiselect` I have to convert the Python inputs to JS, e.g. by serializing them to JSON in Python, and then wrapping the JSON string with browser's `JSON.parse()`, so that the result will be a valid runtime object in the browser:

      ```py
      if input_kwargs.get("selected_items", None) is not None:
          norm_selected_items = [
              SelectOption(value=item, label=item, attrs={})
              if not isinstance(item, SelectOption)
              else item
              for item in input_kwargs["selected_items"]
          ]
          js_props["selected_items"] = json.dumps([
              item._asdict()
              for item in norm_selected_items
          ])
          js_props["selected_items"] = "JSON.parse('" + js_props["selected_items"] + "')"
      ```

   - Thus I also have to guard and raise error if the same kwarg is given as BOTH Python and JS variables.

3. I then return this `js_props` from `get_template_data()`, so I can access it in the template.

4. In the template I use a custom AlpineJS directive `x-props` to be able to pass the JS data from `js_props` to the AlpineJS component by declaring the props in the HTML:

   ```django
    <div
      x-data="multiselect"
      x-props="{
        allItems: {{ js.all_items|escape }},
        selectedItems: {{ js.selected_items|escape }},
        onChange: {{ js.onChange|escape }},
        passthrough: {{ js.passthrough|escape }},
      }"
      class="pt-3 flex flex-col gap-y-3 items-start"
    >
      {% slot "title" / %}

      <template x-for="(item, index) in selectedItems.value" :key="item.value">
        <div class="inline-flex items-center w-full px-2.5 text-sm">
          <span x-html="item.content" class="w-full"></span>
          {% if editable %}
            {% component "Icon"
              name="x-mark"
              attrs:class="ml-1.5 hover:text-gray-400 text-gray-600 cursor-pointer focus:outline-none"
              attrs:@click.stop="removeItem(item)"
            / %}
          {% endif %}
        </div>
      </template>
      ...
   </div>
   ```

5. Thanks to all this, I can access the "props" data in the component's AlpineJS definition as JS variables:

   ```js
    // Define component similarly to defining Vue components
    const Multiselect = AlpineComposition.defineComponent({
      name: "multiselect",
    
      props: {
        /* { label: string, value: string, attrs?: object }[] */
        allItems: { type: Array, required: true },
        selectedItems: { type: Array, required: true },
        passthrough: { type: Object, required: false },
      },
    
      emits: {
        change: (payload) => true,
      },
    
      // Instead of Alpine's init(), use setup()
      // Props are passed down as reactive props, same as in Vue
      // Second argument is the Alpine component instance.
      // Third argument is the reactivity API, equivalent to `@vue/reactivity`
      setup(props, vm, { computed, ref, watch }) {
        // Variables
        const allItems = ref([]);
        const selectedItems = ref([]);
        const items = ref([]);
    
        // Computed
        const allItemsByValue = computed(() => {
          return allItems.value.reduce((acc, item) => {
            acc[item.value] = item;
            return acc;
          }, {});
        });
    
        // Set the initial state from HTML
        if (props.allItems) {
          allItems.value = props.allItems;
        }
    
        if (props.selectedItems) {
          selectedItems.value = props.selectedItems;
        }
    
        // Watch for changes
        watch(selectedItems, () => {
          onItemsChange();
        }, { immediate: true });
    
        // Methods
        const addItem = () => {
          const availableItems = getAvailableItems();
          if (!availableItems.length) return;
    
          // Add item by removing it from available items
          const nextValue = availableItems.shift();
          const newSelectedItems = [
            ...selectedItems.value,
            nextValue,
          ];
    
          // And add it to the selected items
          selectedItems.value = newSelectedItems;
        }
    
        // ...

        return {
          items,
          allItems,
          addItem,
          ...(props.passthrough || {}),
        };
      },
    });
    
    document.addEventListener('alpine:init', () => {
      AlpineComposition.registerComponent(Alpine, Multiselect);
    });
   ```

## Alpine integration implementatio

How I imagine that the integration could simplify this:

1. Since the same kwarg may be given as Python or as JS variables, this AlpineJS integration should hook into `on_component_input()` extension hook and validate that there are no conflicting kwargs that are given both as `my_var` and `js.my_var`.

2. Instead of using `get_template_data()` to generate the data for the AlpineJS component as a`js_props` kwarg, the Alpine component data could be returned directly from `get_js_data()`.

   For extra convenience, we could allow people to return either python or JS data from `get_js_data()`. JS data could be wrapped in `JS()` class or `js()` function to indicate that the string should be treated as a JS code snippet.

   ```py
   class Multiselect(Component):
       class Kwargs(NamedTuple):
           selected_items: list | None = None
           js: MultiselectJsProps | None = None

       def get_js_data(self, args, kwargs: Kwargs, slots, context):
           if kwargs.selected_items:
               selected_items = [
                   SelectOption(value=item, label=item, attrs={})
                   if not isinstance(item, SelectOption)
                   else item
                   for item in input_kwargs["selected_items"]
               ]
           elif kwargs.js.get("selected_items"):
               selected_items = js(kwargs.js)
           else:
               raise ValueError("Missing required kwarg 'selected_items'")

           return {
               "selected_items": selected_items,
               "callback": js("() => console.log('Hello!')"),
           }
   ```

3. Behind the scenes, at `on_component_data()` extension hook, we would post-process the data returned from `get_js_data()`.

   For each key-value pair:

   - If the value is marked as `js()`, then we leave it as is.
   - If NOT, then we serialize it to JSON and wrap in `JSON.parse()` as mentioned earlier.

4. Instead of me having to explicitly set `x-data` and `x-props` the extension could do it itself. 

   So instead of 

   ```django
    <div
      x-data="multiselect"
      x-props="{
        allItems: {{ js.all_items|escape }},
        selectedItems: {{ js.selected_items|escape }},
        onChange: {{ js.onChange|escape }},
        passthrough: {{ js.passthrough|escape }},
      }"
      class="pt-3 flex flex-col gap-y-3 items-start"
    >
   ```

   I could just write

   ```django
    <div
      class="pt-3 flex flex-col gap-y-3 items-start"
    >
   ```

    This would be possible because, at the end of the day, the `x-data` and `x-props` are both HTML attributes, AND applied only to the root HTML element(s).

    We already have logic that exactly parses the rendered HTML and inserts HTML attributes to root element(s). We use it to insert ID attributes like `data-djc-id-cxxxxxx`. It's `set_html_attributes()` that's called from `set_component_attrs_for_js_and_css()`

    So what we would need for this to work is to allow extensions to add extra HTML attributes that should be inserted into the root element(s). This hook would need to access the template / JS / CSS data, so the options are:
    - Either reuse `on_component_data` hook, e.g. passing in an extra dictionary that extensions could populate with the extra root HTML attributes
    - Or define a separate hook that would be solely for populating the HTML attributes.

5. Not 100% sure, but I think this could be also be further extended to allow users to customize where the root of the Alpine component should be, e.g. by setting a `data-djc-js` HTML attribute somewhere in the HTML:

   ```django
    <div class="pt-3 flex flex-col gap-y-3 items-start">
      <div data-djc-js>
        Nested
      </div>
    </div>
   ```

   Which would be then turned into:

   ```django
    <div class="pt-3 flex flex-col gap-y-3 items-start">
      <div x-data="multiselect" x-props="{...}">
        Nested
      </div>
    </div>
   ```

6. The Alpine components created with `alpine-composition` are isolated - e.g. when you set `<div x-data="my_comp">`, then the HTML elements nested within this `<div>` element will NOT be able to access AlpineJS variables from outside.

   This currently is problematic with our slots. This was outlined in the original post for this thread, so I won't go into it. [See here](https://github.com/django-components/django-components/issues/644#issue-2504603637).

   To currently work around it, I had to pass a `passthrough` JS variable which defines alpineJS variables that should be accessible from within slots, and it looks like this:

   When calling `{% component %}` tag:

   ```django
    {% component "Multiselect"
      selected_items=selected_items
      ...
      js:passthrough="{
        assignees: assignees,
        isRoleInputShown: isRoleInputShown,
        getRole: getRole,
        setRole: setRole,
      }"
      attrs:class="pb-8"
    %}
   ```

   Inside component template:

   ```django
    <div
      x-data="multiselect"
      x-props="{
        // ...
        passthrough: {{ js.passthrough|escape }},
      }"
    >
   ```

    Inside Alpine component definition:

    ```js
    const Multiselect = AlpineComposition.defineComponent({
      name: "multiselect",
      props: {
        // ...
        passthrough: { type: Object, required: false },
      },
      setup(props, vm, { computed, ref, watch }) {
        return {
          items,
          // ...,
          ...(props.passthrough || {}),
        };
      },
    });
    ```

7. Lastly, we could get rid of some of the boilerplate from the JS file:

    If the JS file was loaded as a module, then the component object could be exported with `export default`. Thus this would be the way to declare which object in the JS file is the actual component definition. So the same way as Vue does it.

   We could also get rid of this:

   ```js
    document.addEventListener('alpine:init', () => {
      AlpineComposition.registerComponent(Alpine, Multiselect);
    });
   ```

   Because there would a be script that would import all these component modules, and for each module would run the registration script.
    
   Moreover, the component definition would not have the `name` field, because that would be set from the Python component (`class_id`).

   So the JS definition could be simplified to:

    ```js
    export default {
      props: {
        // ...
      },
      emits: {
        // ...
      },
      setup(props, vm, { computed, ref, watch }) {
        return {
          // ...,
        };
      },
    });
   ```

   To implement this, we would probably want to:

    1. During rendering, capture all components that were rendered
        - We already do something like that when collecting the JS / CSS dependencies. But right now we don't allow extensions to hook into that. (see `comp_hashes` in `dependencies.py`
    2. Allow extensions to inject extra JS / CSS scripts based on the above. This could be inserted into the `final_script_tags` and `final_css_tags` variables in `dependencies.py`.
    3. Thus, the extension hook for the Alpine integration would:
       - Receive the list of component class IDs
       - Fetch all components with `get_component_by_class_id()` (or we could pass in already Component classes and not the hashes)
       - For each component class check if Alpine integration is enabled for that component with `Component.Alpine.enabled`, and filter out disabled ones.
       - Lastly it would insert a JS script that would import the the component's JS scripts, and register the Alpine component definitions. Roughly like this:

          ```html
          <script type="module">
            // Import the components
            import Table_a1b2c3 from "/components/Table_a1b2c3.js";
            import Button_a1b2c4 from "/components/Button_a1b2c4.js";
            import Card_a1b2c5 from "/components/Button_a1b2c5.js";

            // Set the Alpine component names to match Python Component class IDs
            Table_a1b2c3.name = "Table_a1b2c3"
            Button_a1b2c4.name = "Button_a1b2c4"
            Card_a1b2c5.name = "Card_a1b2c5"

            // Register the components
            document.addEventListener('alpine:init', () => {
              AlpineComposition.registerComponent(Alpine, Table_a1b2c3);
              AlpineComposition.registerComponent(Alpine, Button_a1b2c4);
              AlpineComposition.registerComponent(Alpine, Card_a1b2c5);
            });
          </script>
          ```

**Comment by JuroOravec** [comment]
Note: We should implement this as part of django-components v1, as this will be one of the most complex plugins we have, so it will help us tease out what the API should be around allowing plugins to 1) modify rendered HTML, 2) modify slot HTML, 3) pass data from component to inside its slots, 4) modify what JS / CSS variables are exposed or doing something with those JS variables, and more.

**Comment by JuroOravec** [comment]
Closing this one, this is implemented in [Citry](https://citry.dev/) ([citry-dev/citry](https://github.com/citry-dev/citry)), the successor to django-components.

Citry integrates with AlpineJS a lot tighter:
- Alpine expressions have the same scoping rules as the Python slot variables
- Python variables returned from `js_data()` are automatically turned into Alpine variables
- Alpine expressions are linted for type errors

This was actually a lot harder than expected. In Citry, each rendered component and slot is wrapped in HTML comments (`<!-- -->`), to mark where each section starts/ends, and Alpine scope is switched based on that. So in the end Alpine expressions have the same strict isolation rules as Python slots (same as Vue or React).

So I definitely won't be backporting this feature to django-components, instead use Citry.

For migration from django-components to Citry, see https://github.com/joeyjurjens/citry-django-poc
